drivers/rtc/rtc_pcf8523.c

```diff
@@ -0,0 +1,966 @@
+		}
+#endif /* CONFIG_RTC_UPDATE */
+
+		control_2 |= PCF8523_CONTROL_2_CTAF | PCF8523_CONTROL_2_CTBF;
+
+		err = pcf8523_write_reg8(dev, PCF8523_CONTROL_2, control_2);
+		if (err != 0) {
+			goto unlock;
+		}
+
+		/* Check if interrupt occurred between CONTROL_2 read/write */
+		err = pcf8523_read_reg8(dev, PCF8523_CONTROL_2, &control_2);
+		if (err != 0) {
+			goto unlock;
+		}
+
+		if (((control_2 & PCF8523_CONTROL_2_AF) != 0U && alarm_callback != NULL) ||
+		    ((control_2 & PCF8523_CONTROL_2_SF) != 0U)) {
+			/*
+			 * Another interrupt occurred while servicing this one, process current
+			 * callback(s) and yield.
+			 */
+			k_sem_give(&data->int1_sem);
+		}
+
+unlock:
+		k_mutex_unlock(&data->lock);
+
+		if (alarm_callback != NULL) {
+			alarm_callback(dev, 0U, alarm_user_data);
+			alarm_callback = NULL;
+		}
+
+		if (update_callback != NULL) {
+			update_callback(dev, update_user_data);
+			update_callback = NULL;
+		}
+	}
+}
+
+static void pcf8523_int1_callback_handler(const struct device *port, struct gpio_callback *cb,
+					  gpio_port_pins_t pins)
+{
+	struct pcf8523_data *data = CONTAINER_OF(cb, struct pcf8523_data, int1_callback);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+
+	k_sem_give(&data->int1_sem);
+}
+#endif /* PCF8523_INT1_GPIOS_IN_USE */
+
+static int pcf8523_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct pcf8523_data *data = dev->data;
+	uint8_t regs[7];
+	int err;
+
+	if (timeptr->tm_year < PCF8523_YEARS_OFFSET ||
+	    timeptr->tm_year > PCF8523_YEARS_OFFSET + 99) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	/* Freeze the time circuits */
+	err = pcf8523_write_stop_bit_unlocked(dev, true);
+	if (err != 0) {
+		goto unlock;
+	}
+
+	LOG_DBG("set time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+		timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+	regs[0] = bin2bcd(timeptr->tm_sec) & PCF8523_SECONDS_MASK;
+	regs[1] = bin2bcd(timeptr->tm_min) & PCF8523_MINUTES_MASK;
+	regs[2] = bin2bcd(timeptr->tm_hour) & PCF8523_HOURS_24H_MASK;
+	regs[3] = bin2bcd(timeptr->tm_mday) & PCF8523_DAYS_MASK;
+	regs[4] = bin2bcd(timeptr->tm_wday) & PCF8523_WEEKDAYS_MASK;
+	regs[5] = bin2bcd(timeptr->tm_mon + PCF8523_MONTHS_OFFSET) & PCF8523_MONTHS_MASK;
+	regs[6] = bin2bcd(timeptr->tm_year - PCF8523_YEARS_OFFSET) & PCF8523_YEARS_MASK;
+
+	/* Write registers PCF8523_SECONDS through PCF8523_YEARS */
+	err = pcf8523_write_regs(dev, PCF8523_SECONDS, &regs, sizeof(regs));
+	if (err != 0) {
+		goto unlock;
+	}
+
+	/* Unfreeze time circuits */
+	err = pcf8523_write_stop_bit_unlocked(dev, false);
+	if (err != 0) {
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int pcf8523_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	uint8_t regs[10];
+	int err;
+
+	/* Read registers PCF8523_CONTROL_1 through PCF8523_YEARS */
```

Comment:
Mutex not needed for these functions?

Reply:
Correct. The registers are read in one go (one I2C transaction) which ensures register coherency due to built-in logic. The mutex is only used for read-modify-write operations.